dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/video_file.h"
+#include "dali/core/error_handling.h"
+
+namespace dali {
+
+void VideoFileCPU::InitAvState() {
+  codec_ctx_ = avcodec_alloc_context3(codec_);
```

Comment:
I think these raw pointers should be wrapped into `unique_ptr` with custom deleters.

Reply:
Unfortunately, sometimes libav API wants to own these pointers. This would require to do `.realease()` and things like that, which I think is even more error prone.
I decided to extract all of the libav related pointers, make a `AVState` structure and move the resource deletion to its destructor. Then I did `unique_ptr<AVState>` field in `VideoFile` to keep everything under control.
I think this is a good compromise between what `libav` wants to do and what is the C++ way of doing things.